### ISSUE TITLE
Using custom ROS2 message types

### ISSUE THREAD

**Issue by zholtovski** [DynoRobotics/unity_ros2#11]
First of all, thanks again for the awesome project! It works like a charm with the provided precompiled libraries, but I'm having trouble with using custom messages.

Here is how I built the custom messages to use in Unity (I'll take *gps_msgs* as an example):
- Using adamdbrw's workspace, I copied the *gps_msgs* package to his workspace and built the whole project.
- Copied the generated shared library files to the *Plugins/Linux/x86_64* directory.
- Copied the generated .dll in the *Plugins* directory

At this point, even my IDE (JetBrains Rider) found the libraries and all symbols of the custom message. There are no compilation errors and the Unity app is running in the editor, but unfortunately fails in the **SpinSome()** method. The error looks like this:

> ArgumentNullException: Value cannot be null.
> Parameter name: src
> System.Runtime.InteropServices.Marshal.Copy (System.IntPtr source, System.Int32[] destination, System.Int32 startIndex, System.Int32 length) (at <567df3e0919241ba98db88bec4c6696f>:0)
> gps_msgs.msg.GPSStatus.ReadNativeMessage (System.IntPtr handle) (at <0d5c52a6f4304631b7bea5ad6e1d01c9>:0)
> gps_msgs.msg.GPSFix.ReadNativeMessage (System.IntPtr handle) (at <0d5c52a6f4304631b7bea5ad6e1d01c9>:0)
> gps_msgs.msg.GPSFix.ReadNativeMessage () (at <0d5c52a6f4304631b7bea5ad6e1d01c9>:0)
> rclcs.Subscription`1[T].TriggerCallback (ROS2.Interfaces.Message message) (at <daea048d63e148ec99a184203650e51a>:0)
> rclcs.Rclcs.SpinOnce (rclcs.INode node, rclcs.Context context, System.Double timeoutSec) (at <daea048d63e148ec99a184203650e51a>:0)
> MonoBehaviourRosNode.SpinSome () (at Assets/Scripts/ROS/MonoBehaviourRosNode.cs:64)
> GPSListener.Update () (at Assets/Scripts/ROS/GPSListener.cs:65)

I haven't dug deeper to find out what happens behind the scenes, but I suspect that the error at runtime has to do with failing message type introspection.

I saw that adamdbrw's project doesn't generate separate shared libraries for separate message type introspections, like they're found in the prebuilt libraries here, i.e.  I don't get a **gps_msgs_gps_fix__rosidl_typesupport_introspection_c_native.so** (where gps_msgs is the package name and GpsFix is the concrete message type). However, it seems like all of the symbols that would be present in those missing library files are in the **libgps_msgs__rosidl_typesupport_introspection_c.so** file.

So my question is, what am I missing and why do I get this runtime error in the SpinSome() function?

**Comment by samiamlabs** [comment]
Hi @zholtovski,

I just released a new version of this project to the master branch. It should now work without any startup scripts for environment variables or apt-get dependencies. Planning to try to put this up as a free Asset on the Unity Asset Store also :)

I automated most of the stuff needed to generate the Plugins folder for Unity while I was working on this. Some basic instructions/pointers for how to build custom messages are now in the README for the new master branch.

Feel free to ask questions if something in the instructions is hard to follow or something is not working.

**Comment by mikkomcmenamin** [comment]
Hi @samiamlabs. I opened the latest master in Ubuntu 18.04 via Unity Hub, Unity version 2019.3.0f6 but I immediately get hundreds of warnings such as 

"Plugins: Couldn't open Assets/ROS2/Plugins/Linux/x86_64/libaction_msgs__python.so, error: libbuiltin_interfaces__python.so: cannot open shared object file: No such file or directory"

Also an error "UnsatisfiedLinkError: librcl.so" 

I was still able to play the SampleScene and subscribe to ROS messages though with the ListenerExample script!

**Comment by samiamlabs** [comment]
Yea, not quite sure how to get rid of the warnings yet...
It seems to be complaining about missing .so files that are not actually missing. 

Good to hear that it works at least :)

**Comment by zholtovski** [comment]
Hi Sam,

thanks again for the fast reply and sorry that it took me so long to respond.

I just tested the new project, but I'm experiencing some linking issues. After building the custom message types and copying them to the Plugins directory in the Unity project, I still can't get gps_msgs to work:

> UnsatisfiedLinkError: libgps_msgs_gps_fix__rosidl_typesupport_c_native.so
> ROS2.Utils.DllLoadUtilsUnix.LoadLibrary (System.String fileName) (at <f635e0a10af54e25ad204b14f8f3d243>:0)
> gps_msgs.msg.GPSFix..cctor () (at <fb9ac3e9d1354e778208d3d4ae95a713>:0)
> Rethrow as TypeInitializationException: The type initializer for 'gps_msgs.msg.GPSFix' threw an exception.
> System.Reflection.MonoCMethod.InternalInvoke (System.Object obj, System.Object[] parameters) (at <437ba245d8404784b9fbab9b439ac908>:0)
> Rethrow as TargetInvocationException: Exception has been thrown by the target of an invocation.
> System.Reflection.MonoCMethod.InternalInvoke (System.Object obj, System.Object[] parameters) (at <437ba245d8404784b9fbab9b439ac908>:0)
> System.RuntimeType.CreateInstanceMono (System.Boolean nonPublic) (at <437ba245d8404784b9fbab9b439ac908>:0)
> System.RuntimeType.CreateInstanceSlow (System.Boolean publicOnly, System.Boolean skipCheckThis, System.Boolean fillCache, System.Threading.StackCrawlMark& stackMark) (at <437ba245d8404784b9fbab9b439ac908>:0)
> System.RuntimeType.CreateInstanceDefaultCtor (System.Boolean publicOnly, System.Boolean skipCheckThis, System.Boolean fillCache, System.Threading.StackCrawlMark& stackMark) (at <437ba245d8404784b9fbab9b439ac908>:0)
> System.Activator.CreateInstance[T] () (at <437ba245d8404784b9fbab9b439ac908>:0)
> rclcs.Subscription`1[T]..ctor (System.String topic, rclcs.Node node, System.Action`1[T] callback, rclcs.QualityOfServiceProfile qualityOfServiceProfile) (at <2e68923f6c464cc68a07b80ecb76c56b>:0)
> rclcs.Node.CreateSubscription[T] (System.String topic, System.Action`1[T] callback, rclcs.QualityOfServiceProfile qos) (at <2e68923f6c464cc68a07b80ecb76c56b>:0)
> GPSListener.StartRos () (at Assets/Scripts/GPSListener.cs:44)
> MonoBehaviourRosNode.Awake () (at Assets/ROS2/Scripts/MonoBehaviour/MonoBehaviourRosNode.cs:34)

I started the Unity Editor from a shell, where I've sourced ROS2. I also tried starting it from a "clean" shell and UnityHub. What am I missing?

**Comment by samiamlabs** [comment]
Have you selected "Load on startup" for all the .so files in the "inspector window" in Unity?

**Comment by zholtovski** [comment]
> Have you selected "Load on startup" for all the .so files in the "inspector window" in Unity?

Oh, I haven't! That fixed it! Thanks a bunch!

**Comment by zholtovski** [comment]
Actually I have another problem now. I don't get any errors or warnings anymore, but it seems like the Unity ROS2 app can't communicate with other ROS2 nodes, the subscription callbacks are never fired up and ros2 topic info doesn't see the Unity subscribers. What could the reason be?

**Comment by samiamlabs** [comment]
Are you using docker? Cyclone DDS does not seem to work for nodes running in containers without network_mode: host

**Comment by zholtovski** [comment]
> Are you using docker? Cyclone DDS does not seem to work for nodes running in containers without network_mode: host

No, I'm not using Docker. I just removed the docker0 network interface to be on the safe side, but that didn't help.

**Comment by samiamlabs** [comment]
Hmm, strange.

Have not had issues like that myself, but I only use ROS2 in containers to not contaminate my system so have not tried with a local install.

Have you tried switching to Cyclone for the other nodes? `export RMW_IMPLEMENTATION=rmw_cyclonedds_cpp`

**Comment by zholtovski** [comment]
> Have not had issues like that myself, but I only use ROS2 in containers to not contaminate my system so have not tried with a local install.
> 
> Have you tried switching to Cyclone for the other nodes? `export RMW_IMPLEMENTATION=rmw_cyclonedds_cpp`

I just tried it, but Cyclone chooses the wrong network adapter for some reason. Do you know if there's a way to specify the network adapter via an environment variable or a config file?

Since I didn't get any network problems with Fast-RTPS, I was wondering if you had a Docker image that supports custom message generation for Fast-RTPS.

**Comment by samiamlabs** [comment]
You should be able to specify network adapter in a file according to https://github.com/eclipse-cyclonedds/cyclonedds#configuration

I was only able to make the linking in Unity to work by building ROS2 with only one DDS implementation and preventing it from using dynamic linking with Poco.
I chose Cyclone because I could make it work without any apt-get system dependencies. Was not able to figure out how to to that for Fast-RTPS.
There are also some modifications needed to the script that copies libraries to the Plugins folder for Fast-RTPS to work.

So short answer, no.

**Comment by zholtovski** [comment]
Hi Sam,

thanks a lot! I tried running all nodes with Cyclone, but I'm facing node discoverability problems, which are obviously not related to UnityRos2.

Coming back to my original question, can I use the older version with the script populating the LD_LIBRARY_PATH environment variable? I would then be able to continue using Fast-RTPS, but why don't I get the introspection .so files?
